### PR TITLE
[2/n] [reconfigurator] handle empty zones/disks use cases within ExampleSystem

### PR DESCRIPTION
**Pull request by sunshowers**
Currently, we build example systems and then mutate the inputs in various ways,
usually to empty out zones and disks.

With this PR, more of these cases are handled within the example system via a builder pattern. This simplifies upcoming work to improve `ExampleSystem` instances for more robust tests.

This also makes a change to `ExampleSystem` to not carry around the blueprint -- logically, the blueprint evolves separately from the underlying system. This change also simplifies some upcoming work around making the collection and input behave correctly.

### CODE REVIEW

File: nexus/reconfigurator/planning/src/planner.rs [hunk 1]
```diff
@@ -1010,11 +1008,17 @@ mod test {
         static TEST_NAME: &str = "planner_add_multiple_nexus_to_one_sled";
         let logctx = test_setup_log(TEST_NAME);
 
-        // Use our example system as a starting point, but strip it down to just
-        // one sled.
+        // Use our example system as a starting point, but strip it down to
+        // just one sled.
+        //
+        // An alternative here would be to pass in nsleds = 1 rather than
```
Comment (andrewjstone) [inline nexus/reconfigurator/planning/src/planner.rs:1014]: Great comment (and pattern to follow).

File: nexus/reconfigurator/planning/src/example.rs [hunk 2]
```diff
@@ -146,29 +235,15 @@ impl ExampleSystem {
                 .unwrap();
         }
 
-        ExampleSystem {
+        // Previously ExampleSystem itself used to carry around the blueprint,
```
Comment (andrewjstone) [inline nexus/reconfigurator/planning/src/example.rs:238]: I don't think we should be referring to what prior versions of the code did in comments. It adds extra cognitive overhead for the reader. John left a similar comment on a PR of mine recently after a chat with Dave regarding comments. 

However, the rest of the comment is useful, so how about something like:

> The blueprint evolves separately from the system -- so it's returned as a separate value.

File: nexus/reconfigurator/planning/src/example.rs [hunk 3]
```diff
@@ -34,17 +33,100 @@ pub struct ExampleSystem {
...
+    (example.collection, example.input, blueprint)
+}
+
+/// A builder for the example system.
+#[derive(Debug, Clone)]
+pub struct ExampleSystemBuilder {
+    log: slog::Logger,
+    test_name: String,
+    nsleds: usize,
+    no_zones: bool,
```
Comment (andrewjstone) [inline nexus/reconfigurator/planning/src/example.rs:55]: Nit: I think negative flags can get confusing and usually prefer to see positive flags so we don't have to use `!no_..` in conditions. How about changing these to `use_zones`, `add_zones`, or `create_zones`, etc...

**Review by andrewjstone — APPROVED** [review]
Looks great. I think separating out the `blueprint` from the `ExampleSystem` is very wise :)

File: nexus/reconfigurator/planning/src/blueprint_builder/builder.rs [hunk 4]
```diff
@@ -2342,7 +2332,15 @@ pub mod test {
             .expect("failed to create builder");
 
             assert!(builder.disks.changed_disks.is_empty());
-            assert!(builder.disks.parent_disks.is_empty());
+            // We expect sleds to be present but not have any disks in them.
+            for (sled_id, disks) in builder.disks.parent_disks {
+                assert_eq!(
+                    disks.disks,
+                    Vec::new(),
+                    "for sled {}, no disks present in parent",
```
Comment (davepacheco) [inline nexus/reconfigurator/planning/src/blueprint_builder/builder.rs:2340]: nitty suggested:

```suggestion
                    "for sled {}, expected no disks present in parent, but found some",
```

Actually, would you get an even more informative message with `assert_eq!(builder.disks.parent_disks, BTreeMap::new())`?

**Review by davepacheco — APPROVED** [review]
This looks much, much better!  Thanks for doing this.

File: nexus/reconfigurator/planning/src/example.rs (hunk 2 above)
Comment (sunshowers) [reply nexus/reconfigurator/planning/src/example.rs:238]: I personally like short notes about battles won in the comments (not just the commit history), but I don't care too much so I'll change it :)

File: nexus/reconfigurator/planning/src/example.rs (hunk 3 above)
Comment (sunshowers) [reply nexus/reconfigurator/planning/src/example.rs:55]: Yeah, that makes sense. Will do.

Comment (sunshowers) [reply nexus/reconfigurator/planning/src/example.rs:55]: Done.

File: nexus/reconfigurator/planning/src/example.rs [hunk 5]
```diff
@@ -34,17 +33,100 @@ pub struct ExampleSystem {
...
+    pub fn no_zones(mut self) -> Self {
+        self.no_zones = true;
+        self
+    }
+
+    /// Do not create any disks in the example system.
+    pub fn no_disks(mut self) -> Self {
+        self.no_disks = true;
+        self
+    }
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply nexus/reconfigurator/planning/src/example.rs:94]: Yeah, I like this -- practically the only values here are 0 and `DEFAULT_NPOOLS`, but API-wise this makes sense.

**Pull request merged**